### PR TITLE
Add search product component

### PR DESCRIPTION
**Pull request by Sampy147**
# Changes made
## Added search page and added this link to the navbar
Some points to note:
- Validation added to FE (keyword cannot be empty or more than 100 chars)
- Only when clicking submit will the search be submitted (clicking enter won't work)
- Searching state will not persist (if you go to another page from the search page and come back/ refresh -> the state will be reset)

## **New change:** Loadmore functionality is integrated with search product page
- After searching in the input, it will display the first 6 products. When loadmore is clicked the next 6 products (or less than 6) will show.
- Everytime the search button is clicked, the search refreshes to show only the first page of searched products
- When there are no more products to show, clicking loadmore will result in "No more products" message
- Loadmore functionality is also robust to refresh/pending requests, in which if a request sent to the BE doesn't work, the functionality should not break (manually tested and verified)

## Deleted searchinput, its test file and the search context
- Search context was deleted as we decided to not persist search state to avoid any additional handling of edge cases, unintended bugs that may occur from state persistance
- Searchinput and its relevant test file was deleted as we realised all search functionality can be migrated onto one test file with very similar functionality. Furthermore, the search page required to track some states such as values and results which were originally in searchInput. Utilising both searchInput in the search page rather than a standalone search page would unnecessarily complicate this access to these key states, and thus both were combined. 
- As searchInput was deleted, its corresponding test file was not necessary

### CODE REVIEW

File: client/src/pages/Search.js [hunk 1]
```diff
@@ -35,13 +135,59 @@ const Search = () => {
...
         </div>
+        <div className="m-2 p-3 d-flex flex-column align-items-center">
+          {
+            // If noMorePage (no new searched products) render this message
+            noMorePage && <h3>No More Products Found</h3>
+          }
+          {
+            // If products list is non-empty and there may still be potential products to be searched,
+            // render loadMore button
+            values.results.length > 0 && !noMorePage && (
```
Comment (jwdavis0200) [inline client/src/pages/Search.js]: can do safety checks values?.results?.length

File: client/src/pages/Search.js [hunk 2]
```diff
@@ -1,12 +1,112 @@
...
+      // Reset key states everytime search button is clicked
+      setPage(1);
+      setNoMorePage(false);
+      setValues({
+        keyword: "",
+        results: [],
+      });
+
+      // Validate keyword cannot be empty
+      if (values.keyword.trim().length === 0) {
```
Comment (jwdavis0200) [inline client/src/pages/Search.js:28]: wouldnt this **almost** always lead to values.keyword.trim().length to be zero since at the start of handleSearch we always call the setValues(...)?

File: client/src/pages/Search.js [hunk 3]
```diff
@@ -1,12 +1,112 @@
...
+        setNoMorePage(true);
+        return;
+      }
+
+      // Relevant key states are updated
+      setPage(newPage);
+      // Adds new searched products to the results array
+      setValues({
+        ...values,
+        results: [...values.results, ...(data?.results || [])],
```
Comment (jwdavis0200) [inline client/src/pages/Search.js]: hmm i think React being async, our Loadmore may give stale data? can use functional form to work around this?

```
setValues((prev) => ({
  ...prev,
  results: [...prev.results, ...(data?.results || [])],
}));
```

**Review by jwdavis0200 — CHANGES_REQUESTED** [review]
Hi @Sampy147 take a look at some of the reviews 👍

File: client/src/pages/Search.js (hunk 2 above)
Comment (Sampy147) [reply client/src/pages/Search.js:28]: Oh yes, noted on this will apply the relevant changes

File: client/src/pages/Search.js (hunk 1 above)
Comment (johnbenedictyan) [reply client/src/pages/Search.js]: It should be values?.results?.length ?? <default value>

But honestly don't like the amount of ? Here

**Review by johnbenedictyan — COMMENTED** [review]
I don't like the number of conditional object access but not a deal breaker HAHAH

**Review by jwdavis0200 — COMMENTED** [review]
With the changes, LGTM! 💯

**Review by jwdavis0200 — APPROVED** [review]
